File: Leetcode/Perfect Squares/Perfect Squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        
        while( n % 4 == 0 ):
            # Reduction by factor of 4
            n //= 4
            
        if n % 8 == 7:
            # Quick response for n = 8k + 7
            return 4
        
        # Check whether n = a^2 + b^2
        for a in range( int(sqrt(n))+1 ):
            
            b = int( sqrt( n - a*a ) )
            
            if ( a**2 + b ** 2 ) == n :
                return (a>0) + (b>0)
            
        # n = a^2 + b^2 + c^2
        return 3

File: Leetcode/Perfect Squares/Perfect_Squares.c
//Problem:https://leetcode.com/problems/perfect-squares/description/

//Approach: I first divide the given number by 4 as often as possible and then do the three-squares check. Dividing by 4 doesn't affect the other checks, and the n % 8 == 7 is cheaper than the prime factorization, so this saves time in cases where we do need four squares.
int numSquares(int n) {
    while (n % 4 == 0)
        n /= 4;
    if (n % 8 == 7)
        return 4;
    for (int a=0; a*a<=n; ++a) {
        int b = sqrt(n - a*a);
        if (a*a + b*b == n)
            return 1 + !!a;
    }
    return 3;
}
